### DISCUSSION TITLE
How to trigger downstream dbt assets for multiple upstream multi_assets with is_required=False?

### DISCUSSION THREAD

**Discussion by TyShkan** [dagster-io/dagster#18667] [Q&A] [answered]
Currently I have a [standard dbt_assets definition](https://docs.dagster.io/_apidocs/libraries/dagster-dbt#dagster_dbt.dbt_assets) and multiple upstream multi_assets with AssetOut(is_required=False).

If I do a full refresh of upstream assets, then dbt_assets trigger as expected.

<img width="221" alt="Screenshot 2023-12-12 at 18 20 10" src="https://github.com/dagster-io/dagster/assets/39046/c222540b-91fb-40a9-b8dd-6e854fecfcc9">

In case if one or more assets of multi_asset don't yield their results, all of the downstream dbt assets would be skipped:
`Skipping step dbt_assets due to skipped dependencies: ['one_of_upstream_multi_asset_name.asset_name']`

<img width="265" alt="Screenshot 2023-12-12 at 17 52 12" src="https://github.com/dagster-io/dagster/assets/39046/1282538f-13d7-4534-8722-e9c29ad0d178">

What could be an approach to keep multi_assets with is_required=False flag and trigger downstream dbt assets as expected?

UPD: [I've tried to #ask-ai](https://dagster.slack.com/archives/C066HKS7EG1/p1702403726410539), but no luck yet

**Comment by sryza** [comment]
@OwenKephart do you know if there's a way to do this? I think this may be a limitation of our current implementatin.

**Reply by OwenKephart** [reply, marked as answer]
@TyShkan Unfortunately this is a limitation of the current implementation -- the core execution logic will never start a downstream step if any of the upstream steps did not return a result.

One potential workaround would be to split your single dbt_assets method into multiple (ideally separated by which assets they're downstream of, although depending on the shape of your graph this may not be possible). This way, if one upstream asset is skipped, only the dbt assets downstream of it would actually be materialized.

**Reply by TyShkan** [reply]
@OwenKephart thanks! I’ve thought about it, in a form of creating assets for each dbt model with individual `select`, but it feels like a sloppy road. What could be a proper way to actually split one dbt_assets method into multiple?

**Reply by OwenKephart** [reply]
I think the best option down that road would (unfortunately) be a sort of factory approach, where you have some function in this general shape:

```python

def get_dbt_assets(manifest) -> Sequence[AssetsDefinition]:

    ret = []
    for select in <some set of dbt selects that break up your graph into smaller pieces>:
        @dbt_assets(manifest=manifest, select=select, name=...)
        def _dbt_assets(...): ...
        
        ret.append(ret)
    return ret

```

**Reply by the4thamigo-uk** [reply]
I have also hit this issue. I think in my case I might have to force a materialization record for all upstream assets, even if nothing was changed.

@OwenKephart I would have thought that as long as the partition has been previously materialized for the upstream asset, then the dbt job should be able to go ahead?

**Reply by emirkmo** [reply]
Is this a limitation of the new `@dbt_assets` approach? I thought this was working fine with the old API?

Anyway, couldn't one create a job that materializes assets even if the upstream assets are partially materialized? Set this job to work on select downstream assets (or make it configurable) and create a sensor to trigger this job whenever one of the watched upstream assets fails to successfully materialize (might need to add some logic to wait until the upstream are fresh). Or somehow based on the freshness of the downstream asset.

**Reply by the4thamigo-uk** [reply]
> Anyway, couldn't one create a job that materializes assets even if the upstream assets are partially materialized?

@emirkmo : something like that might work, but it seems a bit of a hack. 

The logic seems to come from here : https://github.com/dagster-io/dagster/blob/44a4e772e7bd0948f88f7f818b5d7d40a40f6b8e/python_modules/dagster/dagster/_core/execution/plan/active.py#L411
https://github.com/dagster-io/dagster/blob/44a4e772e7bd0948f88f7f818b5d7d40a40f6b8e/python_modules/dagster/dagster/_core/execution/plan/active.py#L244-L245
https://github.com/dagster-io/dagster/blob/44a4e772e7bd0948f88f7f818b5d7d40a40f6b8e/python_modules/dagster/dagster/_core/execution/plan/active.py#L206-L211

Seems that it isnt dbt-specific either. @OwenKephart Is this expected behaviour?

**Reply by the4thamigo-uk** [reply]
@TyShkan Did you find a good workaround for this issue?

**Reply by TyShkan** [reply]
@the4thamigo-uk no, I didn't. For now I yield a result from @multi_asset even if the asset doesn't return any value, that helps to keep things going.

**Reply by the4thamigo-uk** [reply]
Yeah thats what Im having to do, but it is misleading in the UI.

**Comment by caelan-schneider** [comment]
Also dealing with this issue, but not for DBT assets. Basically I have two multi_assets, an upstream and a downstream, and the upstream has is_required=false for each AssetOut, while the downstream has internal_asset_deps mapped appropriately. But because the downstream still requires every asset from the upstream multi_asset to have been materialized, I can never materialize just a subset of assets.

I'm just going to use a factory pattern for now.

**Comment by the4thamigo-uk** [comment]
I also have hit an issue with skipping materialization, which may or may not be related.

https://github.com/dagster-io/dagster/discussions/19994

**Comment by rexledesma** [comment]
We're tracking this feature request here now: https://github.com/dagster-io/dagster/issues/22242
